### ISSUE TITLE
Fatal error when starting php on windows

### ISSUE THREAD

**Issue by madhon** [krakjoe/apcu#223]
C:\php>php --version
PHP Fatal error:  PHP Startup: apc_shm_attach: shmat failed: in Unknown on line
0

The ini file just has the following
;extension=php_apcu.dll
apc.enabled=1
apc.shm_size=32M
apc.ttl=7200
apc.enable_cli=1

This is with PHP 7.0.14 NTS x86 VC14 on Windows 2012

**Comment by Cy4n1d3** [comment]
Happens for me aswell, PHP 7.0.16 NTS x64 on Windows 2008 R2.

**Comment by Fludimir** [comment]
same fo me with PHP 7.1.1 TS x64 on Windows7 x64

**Comment by Jan-E** [comment]
Try loading php_sysvshm.dll before php_apcu.dll:

```
extension=php_sysvshm.dll
extension=php_apcu.dll

```

**Comment by Cy4n1d3** [comment]
Thanks for your input Jan-E, sadly loading `php_sysvshm.dll` still results in this error:

    [20-Feb-2017 12:31:41 Europe/Berlin] PHP Fatal error:  PHP Startup: apc_shm_attach: shmat failed: in Unknown on line 0

**Comment by Jan-E** [comment]
Bugger. Try my builds of php_sysvshm.dll and php_apc.dll:
https://www.apachelounge.com/viewtopic.php?t=6359

I know for sure that they load fine. See for instance
https://phpdev.toolsforresearch.com/php-7.0.16-nts-Win32-VC14-x64.htm

That phpinfo() dump is created by running 'php-cgi.exe phpinfo.php' on the command line with the php.ini that is stored as php.ini.bak now in https://phpdev.toolsforresearch.com/php-7.0.16-nts-Win32-VC14-x64.zip

If you try to reproduce that exactly, you will probably miss some supporting dll's for extensions like oci, interbase, firebird, postgress, imagick and the like. But you'll get the idea, I guess.

**Comment by Jan-E** [comment]
BTW for @krakjoe I stumbled into this issue, because my hoster moved our production Win2008 R2 VPS to a new platform and I got exactly the same error with PHP 5.6.30 x86 NTS and apcu 4.0.11.

After the initial attach error the subsequent ones were like:

> [20-Feb-2017 11:19:17 Europe/Amsterdam] PHP Fatal error:  PHP Startup: apc_shm_create: shmget(0, 33554432, 658) failed: No such file or directory. It is possible that the chosen SHM segment size is higher than the operation system allows. Linux has usually a default limit of 32MB per segment.

I asked them what was changed in their memory management and am still waiting for an answer.

**Comment by Jan-E** [comment]
If you are running under mod_fcgid with Opcache enabled: try to disable Opcache temporarily until all running php-cgi.exe's and all caches are gone. Then enable php_sysvshm.dll and php_apcu.dll. If the pair does not fails fatally, try to enable Opcache once again.

Something of this seems to have solved my problems as well.

**Comment by Cy4n1d3** [comment]
I'll see if I can sneak in a reboot this night.
Disabling Opcache and testing with sysvshm + apcu only still led to the same fatal error (I am already using your builds). Hope I have some more time for investigation later on.

**Comment by tip2tail** [comment]
Hi,

I am also seeing this on windows.  My system details:

PHP Version:		5.6.30
System:			Windows NT SERVER 10.0 build 15025 (Windows 10) AMD64 
Build Date:		Jan 18 2017 19:42:36
Compiler:			MSVC11 (Visual C++ 2012)
Architecture:		x64 
APCu Download:	php_apcu-4.0.11-5.6-nts-vc11-x64.zip
Command:		php -S localhost:8000

If I can help any more please just ask :)

**Comment by The-Don-Himself** [comment]
Same issue here, wamp 64-bit, tried everything above, the only thing is to sadly disable apcu for now.

**Comment by subversivo58** [comment]
The same error here:

      PHP version:  7.1.x NTS 64bit
      OS:                 Windows 8 64bit
      MSVC:            14 (Visual C++ 2015) 64bit
      Apcu:              php_apcu-5.1.8-7.1-nts-vc14-x64

Init PHP crash PHP/FastCGI

**Comment by Jan-E** [comment]
Try something really silly: load a non-existent extension right after apcu. Complete php.ini:
```
log_errors = On
error_log = php_errors.log
extension_dir = "ext"
date.timezone = Europe/Amsterdam
extension=php_sysvshm.dll
extension=php_apcu.dll
extension=php_xcache.dll
```
Result:
```
C:\phpdev\php71.x64>php-cgi -m
[PHP Modules]
apcu
bcmath
calendar
cgi-fcgi
Core
ctype
date
dom
filter
hash
iconv
json
libxml
mcrypt
mysqlnd
pcre
PDO
Phar
readline
Reflection
session
SimpleXML
SPL
standard
sysvshm
tokenizer
wddx
xml
xmlreader
xmlwriter
zip
zlib

[Zend Modules]


C:\phpdev\php71.x64>type php_errors.log
[22-Mar-2017 00:02:46 UTC] PHP Warning:  PHP Startup: Unable to load dynamic library 'ext\php_xcache.dll' - The specified module could not be found. in Unknown on line 0
```
If you remove the last line in the php.ini 'php-cgi -m' crashes with:
```
[22-Mar-2017 00:09:34 UTC] PHP Fatal error:  PHP Startup: apc_shm_attach: shmat
failed: in Unknown on line 0
```
The same happens if you remove the firest two lines in the php.ini (about the error log) !!

Strange error. It looks like shmat does not fail if something is written to php_errors.log. It only happens in the x64-versions of php.

**Comment by Jan-E** [comment]
Additional info: php_sysvshm.dll is not needed, so it seems. This is the minimum php.ini that loads apcu without crashing:
```
log_errors = On
error_log = php_errors.log
extension_dir = "ext"
extension=php_apcu.dll
extension=whatever
```

**Comment by The-Don-Himself** [comment]
This is a bit weird in my view but I can confirm @Jan-E workaround hasn't crashed my Wampserver 64bit since I added the non-existing extension. Will promptly update the thread if it does. Thanks.

**Comment by Jan-E** [comment]
@krakjoe and/or @weltling It is an unhandled exception in ntdll.dll with the following call stack:

![apcu](https://cloud.githubusercontent.com/assets/1757825/24178133/74a39d24-0ea7-11e7-82d0-cd6bb8aee515.png)

The error is above my head, but maybe this gives a clue.

**Comment by Jan-E** [comment]
![apcu2](https://cloud.githubusercontent.com/assets/1757825/24178608/689620ee-0eaa-11e7-9e3d-74911f0c02ce.png)

**Comment by Jan-E** [comment]
https://github.com/krakjoe/apcu/blob/master/apc_lock.c#L207
calls
https://github.com/krakjoe/apcu/blob/master/apc_windows_srwlock_kernel.c#L61

**Comment by Jan-E** [comment]
https://github.com/krakjoe/apcu/issues/223#issuecomment-288277324:
> Functions introduced by this commit: https://github.com/krakjoe/apcu/commit/54bbe072c89e42c5070e996e4d6236fcd135e72e

This commit with the title "integrated srwlock kernel from APC" includes a comment:
> /* XXX kernel lock mode only for now, compatible through all the wins, add more ifdefs for others */

'Compatible through all the wins' does not seem to be true anymore, now that the x64 build for PHP7 causes troubles on Win7 and Windows Server 2008.

Besides that, I remembered that the original APC had two choices: srwlock kernel and srwlock native. The commit that introduced this in the original APC explicitly stated that srwlock native was the best option for Win7+:
http://git.php.net/?p=pecl/caching/apc.git;a=commit;h=b257e7f6cb1d139e11a6be4b955db7b049588841

Maybe the native srwlock should be used in the apcu extension as well.

**Comment by lucaspiris** [comment]
Hi,

I have same problem with php 7.1.8 and apcu 5.1.8:

Fatal error: PHP Startup: apc_shm_attach: shmat failed: in Unknown on line 0

extension=php_apcu.dll
apc.enabled=1
apc.shm_segments=1
apc.shm_size=32M
apc.ttl=7200
apc.enable_cli=1
apc.serializer=php

Anyone solve this?

Regards
Lucas

**Comment by Wirone** [comment]
Same here, PHP 7.1.4 x86 TS + APCu 5.1.8 + APCu-BC 1.0.3

```
[APCu]
; http://windows.php.net/downloads/pecl/releases/apcu/
extension=php_apcu.dll
; https://pecl.php.net/package/apcu_bc/1.0.3/windows
extension=php_apcu_bc-1.0.3-7.1-ts-vc14-x86.dll

apc.enabled=1
apc.shm_segments=1
apc.shm_size=32M
apc.ttl=7200
apc.enable_cli=1
apc.serializer=php
```

**Comment by Jan-E** [comment]
@Wirone: could you try the php_apcu.dll in https://phpdev.toolsforresearch.com/php-7.1.4-Win32-VC14-x86.zip without the 'extension=whatever' in your php.ini?

I am rebuilding all other PHP 7.0.18 and PHP 7.1.4 builds at https://www.apachelounge.com/viewtopic.php?t=6359 at the moment with the patch in https://github.com/krakjoe/apcu/issues/244

That patch writes an E_WARNING to the error_log before initializing the shared memory segments.

**Comment by Wirone** [comment]
@Jan-E I've used dll from mentioned ZIP and I got a lot of `Unable to load dynamic library` errors because all extensions were loaded from `C:\php\ext\` but it's NOT my installation dir. Then I've set `extension_dir` to `ext` (without path) and restarted Apache. CLI output:

```
-> % php --version
apcu.shm_segments will be initialized
PHP 7.1.4 (cli) (built: Apr 11 2017 20:08:12) ( ZTS MSVC14 (Visual C++ 2015) x86 )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
```

Warning is gone, but APC is not mentioned in my phpinfo page (the same goes for cURL which is enabled in `php.ini` but not available - `curl_init` throws `UndefinedFunctionException`).

I've installed 7.1.4 today, before I was using 7.1.0 and everything was ok.

**Comment by Wirone** [comment]
I've downloaded PHP 7.1.4 once again, exctracted clean dir, copied `php.ini` from 7.1.0, copied working APCu DLLs form 7.1.0 dir (previously I've downloaded newest DLL for APCu and APCu-BC)... and it works.

**Comment by Jan-E** [comment]
Do you remember where you downloaded the php_apc*.dll's fir PHP 7.1.0?

In the mean time I will check the ones from https://phpdev.toolsforresearch.com/php-7.1.0-Win32-VC14-x86.zip

**Comment by Jan-E** [comment]
Do they mention a version in the phpinfo output? Then you could check http://windows.php.net/downloads/pecl/releases/apcu/ to see if they were there.

**Comment by weltling** [comment]
I was trying to reproduce this, but have no crash so far, current PHP-7.1 and APCu dev. I'd ask for further info on how to reproduce the crash, maybe there's some specific ini, system config, etc. required.

Thanks.

**Comment by Jan-E** [comment]
Mail me the IP-address from where you are connecting and I will give you RDP access to a Win7 virtualbox which has the crash.

**Comment by Jan-E** [comment]
@weltling On the internals-win mailing list I already posted a screenshot of the crash on that Win7 box, but here it is once again: https://phpdev.toolsforresearch.com/apcu.png

**Comment by Wirone** [comment]
@Jan-E My DLLs for 7.1.0 were downloaded from:
http://windows.php.net/downloads/pecl/releases/apcu/
https://pecl.php.net/package/apcu_bc/1.0.3/windows

```
apc
APC Compatibility => 1.0.3
APCu Version => 5.1.2
Build Date => Dec  2 2016 15:27:12

apcu
APCu Support => Enabled
Version => 5.1.8
APCu Debugging => Disabled
MMAP Support => Disabled
Serialization Support => php
Build Date => Jan 16 2017 08:45:30
```

**Comment by weltling** [comment]
Thanks for the images, @Jan-E, but i'd prefer a way it could be reproduced somewhere one can debug.

From the thread, https://github.com/krakjoe/apcu/issues/223#issuecomment-288271719 seems to be due to something else, the fatal error is what is originally reporting. Maybe there's something useful in the event log after this fatal happened?

Thanks.

**Comment by Jan-E** [comment]
> Thanks for the images, @Jan-E, but i'd prefer a way it could be reproduced somewhere one can debug.

On that Virtualbox you can do whatever you want. I will put back the current snapshot afterwards.

**Comment by weltling** [comment]
@Jan-E could you check the latest master please? I guess your BT might still be a side effect witnessing a core bug, not APCu. I just still couldn't reproduce this completely locally. 

Thanks for the offer otherwise, but it's really something taking too much time to setup a debug environment, especially on VMs where huge installs like VS take times.

Thanks.

**Comment by Jan-E** [comment]
@weltling Took a while, because I was busy with the other half of my life: https://www.albatros-amsterdam.nl/

Anyway, it looks like you've solved the issue. I also ran the tests with PHP 7.1.4 NTS x64. The only test that failed was apc54_014.phpt:
```
=====================================================================
Number of tests :   33                32
Tests skipped   :    1 (  3.0%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :    1 (  3.0%) (  3.1%)
Expected fail   :    0 (  0.0%) (  0.0%)
Tests passed    :   31 ( 93.9%) ( 96.9%)
---------------------------------------------------------------------
Time taken      :   10 seconds
=====================================================================

=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
APC: Bug #61742 preload_path does not work due to incorrect string length (variant 1) (php
5.4) [/php-sdk/php71dev/ext/apcu/tests/apc54_014.phpt]
=====================================================================
```
Output of apc54_014.phpt:
```
Notice: Undefined offset: 1 in N:\php-sdk\php71dev\ext\apcu\tests\server_test.inc on line 101
bool(true)
string(3) "123"
bool(true)
string(3) "123"

Warning: fsockopen(): unable to connect to 127.0.0.1:8871 (No connection could be made because the target machine actively refused it.
) in N:\php-sdk\php71dev\ext\apcu\tests\server_test.inc on line 144
connect failed errno=10061 errstr='No connection could be made because the target machine actively refused it.
'``` 
Running apc54_014.php gave the expected output, so it is probably something in the way I ran the tests:
```
N:\php-sdk\php71dev\x64\Release\php-7.1.4>php /php-sdk/php71dev/ext/apcu/tests/apc54_014.ph
p
PHP 7.1.4 Development Server started at Sun Apr 16 05:02:41 2017
Listening on http://127.0.0.1:8871
Document root is N:\php-sdk\php70dev\ext\apcu\tests
Press Ctrl-C to quit.
PHP 7.1.4 Development Server started at Sun Apr 16 05:02:42 2017
Listening on http://127.0.0.1:8872
Document root is N:\php-sdk\php70dev\ext\apcu\tests
Press Ctrl-C to quit.
PHP 7.1.4 Development Server started at Sun Apr 16 05:02:42 2017
Listening on http://127.0.0.1:8873
Document root is N:\php-sdk\php70dev\ext\apcu\tests
Press Ctrl-C to quit.
bool(true)
string(3) "123"
bool(true)
string(3) "123"
bool(true)
string(3) "123"
bool(true)
etc
```

**Comment by weltling** [comment]
Good news, thank's for checks, @Jan-E! Life always comes first, OFC :)

Here are the snapshots for everyone to test http://windows.php.net/downloads/pecl/snaps/apcu/5.1.9dev/ 

Thanks.

**Comment by Jan-E** [comment]
For all those that experience this problem in PHP 7 or PHP 7.1: I have recompiled all my PHP 7 and PHP 7.1 builds at https://www.apachelounge.com/viewtopic.php?t=6359

Try replacing your php_apcu.dll with the one from the appropiate build to see if that solves the problem.

**Comment by Jan-E** [comment]
@anatol I was just writing a comment with the same intention...

Am I correct that the patch can be backported to the PHP5 branch? There are sporadic reports of the same issue in PHP 5.6.

**Comment by weltling** [comment]
Oh, i see. I've cherry-picked a small part into PHP5, snaps are under way http://windows.php.net/downloads/pecl/snaps/apcu/4.0.11/ . Would unwillingly touch the locking part for PHP5, so not done for now.

Thanks.

**Comment by MDT-FGI** [comment]
5.1.9dev fixes the issue for me. Thanks.

**Comment by hakuno** [comment]
@weltling @Jan-E 

Same error. The 5.1.9dev has no effect for me. How do I solve? I have enabled both apcu and sysvshm on the IIS (PHP 7.0.21 x64).

**Comment by VSG24** [comment]
@weltling When can we expect a stable release?

**Comment by Jan-E** [comment]
@VSG24 No release, but the apcu.dll's in my builds are quite stable:
https://www.apachelounge.com/viewtopic.php?t=6359
They are built using the latest sources.

**Comment by VSG24** [comment]
@Jan-E What do you mean "No release"?? If it's stable why not remove "dev" from 5.1.9??

**Comment by Jan-E** [comment]
I cannot add a release in this repo or PECL. As soon as @krakjoe or @remicollet tags a new release in https://github.com/krakjoe/apcu/releases it will be automagically added in PECL as well.

**Comment by nikic** [comment]
This fix is part of the apcu 5.1.9 release, so I guess this can be closed.
